Stepper: Add Flow.isSignupFlow

daledupreez opened the pull request:
Related to:
* https://github.com/Automattic/automattic-analytics/issues/406
* https://github.com/Automattic/wp-calypso/pull/85769

## Proposed Changes

* This PR explores requiring an explicit declaration for Stepper flows to be considered a signup flow via a new `Flow.isSignupFlow` flag. In turn, this explicit declaration controls whether we trigger a `calypso_signup_start` event for the flow.
  - The driver for this particular change is that after https://github.com/Automattic/wp-calypso/pull/85769 was merged, we're now logging `calypso_signup_start` events for a number of onboarding flows, not only signup flows
* ~The PR also explores a new `useSignupStartEventProps()` hook on the `Flow` type to make it possible for flows to augment the event props for the `calypso_signup_start` event.~ This part of the PR will be tackled in a follow-up. This means we will continue to fire two `calypso_signup_start` events for the `/setup/ecommerce` flow.
* Finally, the PR sets the `isSignupFlow` flag to `true` for a subset of flows that I _think_ should be considered signup flows, primarily for cases where they may create a site.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Run this branch locally or via Calypso.live (as per the link in [this comment](https://github.com/Automattic/wp-calypso/pull/87758#issuecomment-1959015650))
* Ensure you can view Tracks events easily, e.g. via Tracks Vigilante
* Run through various `/setup/*` flows, and verify that a `calypso_signup_start` Tracks event is only fired when the `isSignupFlow` flag is true in the code. Verify that we do not trigger a `calypso_signup_start` event for flows where `isSignupFlow` is false.
  - Note that this is distinct from the `calypso_signup_step_start` event -- we do expect that to be triggered
  - Also note that we expect two `calypso_signup_start` events to be triggered for the `/setup/ecommerce` flow. We will fix that in a follow-up PR.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [N/A] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [N/A] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

daledupreez committed:
Stepper: Add Flow.isSignupFlow and Flow.useSignupStartEventProps()

taipeicoder commented:
@daledupreez the approach LGTM, thanks for the implementation. I'll test the PR tomorrow.

gabrielcaires commented on client/landing/stepper/declarative-flow/ai-assembler.ts:
```diff
@@ -29,6 +29,7 @@ const SiteIntent = Onboard.SiteIntent;
 
 const withAIAssemblerFlow: Flow = {
 	name: AI_ASSEMBLER_FLOW,
+	isSignupFlow: true,
```
Considering this flag is very important, what do you think about updating the type to make it required?

So the flow that the user signup is not required should explicitly set is as false.

taipeicoder replied:
That's a great point! No strong preference here, my two cent is that I would delegate the responsibility of adding the flag to the person that's implementing the new flow. While the flag is important, it's not essential for the flow to function.

markbiek replied:
I have mixed feelings. My first thought is that I agree with @taipeicoder. But all of the existing flows are currently signup flows so it may make more sense to go with @gabrielcaires 's suggestion.

gabrielcaires commented on client/landing/stepper/declarative-flow/internals/index.tsx:
```diff
@@ -159,8 +159,8 @@ export const FlowRenderer: React.FC< { flow: Flow } > = ( { flow } ) => {
-		if ( isFlowStart() ) {
-			recordSignupStart( flow.name, ref );
+		if ( flow.isSignupFlow && isFlowStart() ) {
+			recordSignupStart( flow.name, ref, flow.useSignupStartEventProps?.() ?? {} );
```
Would be better to update the recordSignupStart to have the third param with a default value? 
`recordSignupStart(param1, param2, param3={})`.

The effect is the same but it will make the code describing our usage better.

taipeicoder replied:
Since the `useSignupStartEventProps` defined in `/tailored-ecommerce-flow.ts` calls `useSelect`, Stepper breaks with the message: 

`Uncaught Error: Invalid hook call. Hooks can only be called inside of the body of a function component.`

markbiek replied:
Just noting here as well that I'm removing the `flow.useSignupSstartEventProps` call to unblock this PR. We can iterate on it.

gabrielcaires commented on client/landing/stepper/README.md:
```diff
@@ -38,6 +38,8 @@ To create a flow, you only have to implement `useSteps` and `useStepNavigation`.
 
 There is also an optional `useSideEffect` hook. You can implement this hook to run any side-effects to the flow. You can prefetch information, send track events when something changes, etc...
 
+There is an optional `isSignupFlow` flag that _MUST be implemented for signup flows_ (generally where a new site may be created), as well as an optional `useSignupStartEventProps()` hook. The `isSignupFlow` flag controls whether we'll trigger a `calypso_signup_start` Tracks event when the flow starts, while the `useSignupStartEventProps()` hook makes it possible to add custom field to the event t ...
```
I think is better to document it directly on the type to make it more integrated with the code editors

taipeicoder replied:
Both places make sense to me! My vote goes to adding notes in both places (which this PR already does).

markbiek replied:
I think adding notes in both places will be the most helpful to everyone.

taipeicoder commented on client/landing/stepper/declarative-flow/tailored-ecommerce-flow.ts:
```diff
@@ -48,16 +48,16 @@ function getPlanFromRecurType( recurType: string ) {
-	useSteps() {
-		const recurType = useSelect(
+	isSignupFlow: true,
+	useSignupStartEventProps() {
```
See https://github.com/Automattic/wp-calypso/pull/87758/files#r1500191298.
We could try calling `useSignupStartEventProps` outside the `useEffect` might we might need to add checks to prevent the event triggering multiple time.

For the sake of unblocking this PR, propose we don't implement `useSignupStartEventProps` for now, trigger the event as it previously did, and iterate in a follow-up PR. WDYT?

taipeicoder replied:
@daledupreez @gabrielcaires  WDYT if we move `useSignupStartEventProps` as the following:

```
  const eventProps = flow.useSignupStartEventProps?.() ?? {};                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
  useEffect( () => {                                                                                                                                                                                                                                                                                           
    if ( flow.isSignupFlow && isFlowStart() ) {                                                                                                                                                                                                                                                                  
      recordSignupStart( flow.name, ref, eventProps );                                                                                                                                                                                                                                                     
    }                                                                                                                                                                                                                                                                                                    
  }, [ flow, ref ] );
```

markbiek replied:
@taipeicoder I tried a couple different ways of moving `flow.useSignupStartEventProps` outside of the `useEffect` and didn't have much luck.

I'm going with your suggestion of removing it for this PR and then iterating in a follow-up.

taipeicoder replied:
Yep, sounds good!

daledupreez replied:
Should we also remove the function for this iteration, especially if we're not calling it?

daledupreez replied:
I've removed the function and the type in 5b604fec115b62601e6b795b41c9333ddb9bf927. That way we can name things however we like in the follow-up PR.

taipeicoder commented:
Tested all the flows with `isSignupFlow: true`:

- ai-assembler
- assembler-first-flow
- design-first
- free
- hundred-year-plan
- import-flow
- import-hosted-site
- link-in-bio-tld
- link-in-bio
- new-hosted-site-flow
- newsletter
- reblogging
- tailored-ecommerce-flow
- trial-wooexpress-flow
- videopress

They work as intended except for tailored-ecommerce-flow as described https://github.com/Automattic/wp-calypso/pull/87758#discussion_r1500196769

markbiek committed:
Make isSignupFlow a required prop

markbiek committed:
Remove calls to useSignupStartEventProps from recordSignupStart

There's some refactoring to be done in order for this to work. We'll
tackle that in a future PR.

taipeicoder commented on client/landing/stepper/declarative-flow/internals/types.ts:
```diff
@@ -102,6 +102,12 @@ export type Flow = {
 	classnames?: string | [ string ];
+	/**
+	 * Optional flag to indicate if the flow is a signup flow.
+	 * Thus MUST be implemented for signup flows.
```
[nit] Typo?
```suggestion
	 * This MUST be implemented for signup flows.
```

daledupreez commented on client/landing/stepper/declarative-flow/internals/types.ts:
```diff
@@ -102,6 +102,12 @@ export type Flow = {
 	title?: string;
 	classnames?: string | [ string ];
+	/**
+	 * Optional flag to indicate if the flow is a signup flow.
```
The flag is no longer optional.
```suggestion
	 * Flag to indicate if the flow is a signup flow.
```

daledupreez replied:
Updated in edb04500ed543fb96d408c68cfdef6e77631a7c6.

daledupreez commented on client/landing/stepper/README.md:
```diff
@@ -38,6 +38,8 @@ To create a flow, you only have to implement `useSteps` and `useStepNavigation`.
 
 There is also an optional `useSideEffect` hook. You can implement this hook to run any side-effects to the flow. You can prefetch information, send track events when something changes, etc...
 
+There is an optional `isSignupFlow` flag that _MUST be implemented for signup flows_ (generally where a new site may be created), as well as an optional `useSignupStartEventProps()` hook. The `isSignupFlow` flag controls whether we'll trigger a `calypso_signup_start` Tracks event when the flow starts, while the `useSignupStartEventProps()` hook makes it possible to add custom field to the event t ...
```
With the changes to make `isSignupFlow` required, the wording here should be updated to match.
```suggestion
There is a required `isSignupFlow` flag that _MUST be set to `true` for signup flows_ (generally where a new site may be created), and should be `false` for other flows. There is also an optional `useSignupStartEventProps()` hook. The `isSignupFlow` flag controls whether we'll trigger a `calypso_signup_start` Tracks event when the flow starts, while the `useSignupStartEventProps()` hook makes it possible to add custom fields to the event that will get logged.
```

daledupreez replied:
Updated in edb04500ed543fb96d408c68cfdef6e77631a7c6.

daledupreez commented on client/landing/stepper/declarative-flow/tailored-ecommerce-flow.ts:
```diff
@@ -48,16 +48,16 @@ function getPlanFromRecurType( recurType: string ) {
 			[]
 		);
 
-		useEffect( () => {
```
If we're not going to call `useSignupStartEventProps()` in this PR, might it be cleaner to leave the secondary event logging so we at least keep getting the events with `recur` defined? Otherwise we won't be able to see which flow the user actually went through for pricing etc.

daledupreez replied:
I've gone ahead and removed `useSignupStartEventProps()` from this PR entirely, and left things as they were in the original implementation, duplicate event and all. The change that does this is 5b604fec115b62601e6b795b41c9333ddb9bf927.

daledupreez reviewed (commented):
I think we have some doc cleanup to do, and I think we should probably undo the logic relating to the `tailored-ecommerce-flow` for now so we can tackle that in a follow-up.

daledupreez committed:
Update docs and type info as `isSignupFlow` is required

daledupreez committed:
Remove `useSignupStartEventProps()` from this PR

taipeicoder commented on client/landing/stepper/declarative-flow/tailored-ecommerce-flow.ts:
```diff
@@ -48,6 +48,7 @@ function getPlanFromRecurType( recurType: string ) {
 
 const ecommerceFlow: Flow = {
 	name: ECOMMERCE_FLOW,
+	isSignupFlow: true,
```
`isSignupFlow` should be `false` here since the flow will trigger the event on its own rather than depending on Stepper to do it.

taipeicoder reviewed (approved):
Preemptive approving this. We only need to change the `isSignupFlow` to `false` for the tailored-ecommerce-flow, otherwise the event `calypso_signup_start` will be triggered twice.

daledupreez committed:
Disable isSignupFlow for ecommerce flow to avoid duplicates

ddc22 commented on client/landing/stepper/declarative-flow/internals/index.tsx:
```diff
@@ -159,7 +159,7 @@ export const FlowRenderer: React.FC< { flow: Flow } > = ( { flow } ) => {
 
 	useEffect( () => {
-		if ( isFlowStart() ) {
+		if ( flow.isSignupFlow && isFlowStart() ) {
```
@daledupreez  I wonder if stepper events should have been called `calyps_stepper_start`

CC @alshakero @southp

southp replied:
I think that depends on which conceptual layer these events reside. In this case, we can ask whether we are tracking things about whether a flow is implemented on Stepper or a general signup flow experience. In this case I think we are tracking for the latter, so I think it'd be better to abstract the name away from having anything Stepper specific. If we really want to track implementation specific info, I'd recommend that we use a prop instead of a name.

Does that make sense? :)

ddc22 replied:
I agree with the sentiment, but whish there was a way to track these types of bugs more easily.
